shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1699,7 +1680,8 @@ - (CGRect)caretRectForPosition:(UITextPosition*)position {
     CGRect characterAfterCaret = rects[0].rect;
     // Return a zero-width rectangle along the upstream edge of the character after the caret
     // position.
-    if ([self isRTLAtPosition:index]) {
+    if ([rects[0] isKindOfClass:[FlutterTextSelectionRect class]] &&
```

Comment:
The old implementation doesn't seem to have this check?

Reply:
Also you might have to special case it when `characterAfterCaret` is a newline character,  iirc SkParagraph reports it as an empty textbox at the end of the current line (instead of at the beginning of the next line).

Reply:
Here we use `selectionRectsForRange` instead of directly looking at the array. I put a check, because I guess if someone subclasses and reimplemented `selectionRectsForRange`, this would crash if it didn't return specifically `FlutterTextSelectionRect`.

Not sure what you mean about the special case. But for end-of-line, it does work correctly because I have added `affinity` into the `FlutterTextPosition`. So it will go to the `characterBeforeCaret` section.

Reply:
If the document is "{newline}A" (so 2 characters),  placing the cursor at `(1, upstream)` should point to the start of the second line. I think it would fall into the "remaining cases" category (around line 1718) and returns
```
    return CGRectMake(characterBeforeCaret.origin.x + characterBeforeCaret.size.width,
                      characterBeforeCaret.origin.y, 0, characterBeforeCaret.size.height);
```
which is going to be the right edge of the zero-width TextBox skparagraph returns for the newline character at the first line.

Reply:
As I understand upstream/downstream, it only comes into play where you have word wrap breaking lines. When you have a hard line-break, we have distinct indices at the end of line 1 and the beginning of line 2. And so most of the time the affinity is downstream, exccept at the end of a word-wrapped line.

I tried out the "`\n`A" scenario and it seems to work as expected in all 3 possible cursor locations. In the specific one you highlighted (cursor at index 1), it used the middle branch (`rects.count == 2 && affinity == UITextStorageDirectionForward`). I think this isn't a concern as you will never actually get that affinity `(1, upstream)` using either keyboard or touch?

Reply:
> you will never actually get that affinity (1, upstream) using either keyboard or touch?

Ah that's probably true.

But I think newlines are still special (sorry my example was bad). I tried getting the text boxes for "A\n":
```dart
    final TextPainter painter = TextPainter(
      textDirection: TextDirection.ltr,
      text: const TextSpan(text: 'A\n'),
    )..layout();
    for (int i = 0; i < 2; i++) {
      print(painter.getBoxesForSelection(TextSelection(baseOffset: i, extentOffset: i + 1)));
    }
```
the output:
```
[TextBox.fromLTRBD(0.0, 0.0, 14.0, 14.0, TextDirection.ltr)]
[TextBox.fromLTRBD(14.0, 0.0, 14.0, 14.0, TextDirection.ltr)]
```
when `index == 1` and the affinity is `UITextStorageDirectionForward` this should place the cursor at the start of the second line. But if you look at the output both `TextBox`es are on the first line.

Reply:
Ah I get it now. But I don't know if we can fix that here, if we don't even have any text box on the second line, we can't even make a guess of the correct position? It needs to be fixed upstream in skparagraph. And there isn't any regression in this PR. I hope we can ignore this, and merge it to avoid the hard crash.

Reply:
SGTM. Probably not a SkParagraph issue I think, we should probably add a bit more information to rects the framework sends.